### ISSUE TITLE
git-annex > 10.20220504 breaks `git-clone` from `datalad-annex::...exporttree=yes` (any special remote type)

### ISSUE THREAD

**Issue by mih** [datalad/datalad-next#72]
TL;DR:

git-annex now forcefully and unconditionally distrusts `exporttree=yes` special remotes. This makes it seemingly impossible for the `datalad-annex::` remote helper implementation to use `annex fsck` to probe a specific key as a remote.

----

Mainline fails https://ci.appveyor.com/project/mih/datalad-next/build/job/98jegq2gaw7i04rg

```
eq_(ds.repo.get_hexsha(ds.repo.get_corresponding_branch()),
1386AssertionError: 'b1c9aef80331c775dc06372d5cb2536543a15455' != None
1387
```

Possibly a git version issue. Does not replicate for me locally.

Update: Nope, git version in both cases is 2.36.1

Looking into the failure of `datalad_next.tests.test_create_sibling_webdav.test_common_workflow_export` specifically: this is near-identical to `datalad_next.tests.test_create_sibling_webdav.test_common_workflow_explicit_cred` (which passes), but uses the `--filetree` mode for the webdav sibling. The dataset clone the respective check is failing on, indeed contains no commits.

Rerunning the `clone` call does not produce any messages.

Manually looking at the repository deposited via `push` on webdav (and the underlying filesystem location) reveals no problem -- looks good with all expected content/branches/commits present.

So the issue is clone from a `filetree`-mode deposit.

Appveyor uses a git-annex snapshot `git-annex version: 10.20220505-g2df856370`, locally I have `git-annex version: 10.20220504`

Deploying this snapshot locally also triggers the failure. This points fair and square to a behavior change in git-annex past 10.20220504.

Git-annex did indeed do two export-related changes since that release:

```
  * Special remotes with importtree=yes or exporttree=yes are once again
    treated as untrusted, since files stored in them can be deleted or
    modified at any time.
    (Fixes a reversion in 8.20201129)
  * Special remotes using exporttree=yes and/or importtree=yes now
    checksum content while it is being retrieved, instead of in a separate
    pass at the end.
 ```

The breakage is also detected by the more direct test of the `datalad-annex` remote helper: `datalad_next.gitremote.tests.test_datalad_annex.test_export_remote` also failing to locate any content in a clone, immediately after cloning. Here is a demonstrator:

```
% datalad create --no-annex myds
% cd myds
% git remote add dla 'datalad-annex::file:///tmp/myremote?type=directory&directory={path}&encryption=none&exporttree=yes'
% git push -u dla

# seems all good, remote deposit looks good

% cd ..
% git clone 'datalad-annex::file:///tmp/myremote?type=directory&directory={path}&encryption=none&exporttree=yes' myclone
Cloning into 'myclone'...
warning: You appear to have cloned an empty repository.

# looking at the internals
% git -C .git/dl-repoannex/origin/mirrorrepo log
fatal: your current branch 'master' does not have any commits yet
```

Diving into the code, it seems that the remote helper fails to detect a deposit at the remote. This leads to an empty local mirror repository. The immediate cause of this misdetection is caused by

```
[DEBUG] Finished ['git', '-c', 'diff.ignoreSubmodules=none', 'annex', 'fsck', '-f', 'origin', '--fast', '--key', 'XDLRA--refs', '--debug', '-c', 'annex.dotfiles=true'] with status 1 
```

It exists with code `1` indicating that there would be no such key.

<details>
<summary>
Protocol of a debugging detour that led nowhere
</summary>
However, running the equivalent command again, reports what is to be expected: success.

```
% git -C myclone/.git/dl-repoannex/origin/repoannex annex fsck -f origin --fast --key XDLRA--refs
fsck XDLRA--refs ok
(recording state in git...)
```

This discrepancy does not appear to be caused by a race conditon. Adding a 30s wait before the command does not change the result. At the same time, immediately running the command line equivalent always succeeds.

Matching the command line call and the internal execution more closely with

```
% git -C myclone/.git/dl-repoannex/origin/repoannex -c diff.ignoreSubmodules=none annex fsck -f origin --fast --key XDLRA--refs --debug -c annex.dotfiles=true
```

does not make a difference, also succeeds.

Conclusion: The reason why I did not see that outside that session was simply using an older git-annex in the other shell :(

</details>

Going deeper, here is what that `fsck` call yields internally. I reformatted the output to make things a bit clearer:

```
[DEBUG] Finished ['git', '-c', 'diff.ignoreSubmodules=none', 'annex', 'fsck', '-f', 'origin', '--fast', '--key', 'XDLRA--refs', '--debug', '-c', 'annex.dotfiles=true'] with status 1 
[Level 8] CommandError: 'git -c diff.ignoreSubmodules=none annex fsck -f origin --fast --key XDLRA--refs --debug -c annex.dotfiles=true' failed with exitcode 1 under /home/mih/hacking/datalad/next/tmp/myclone/.git/dl-repoannex/origin/repoannex [out: 'fsck XDLRA--refs (fixing location log) 

  (Avoid this check by running: git annex dead --key )
failed
(recording state in git...)'] [err: '....']
```

The `err` output I cut from the listing is elaborate and is primarily the XDLRA backend process talking:

<details>
<summary>
Importantly, it contains the output:

```
  Only these untrusted locations may have copies of XDLRA--refs
        0cc19d93-2885-44c8-83ba-30c2dd3afe40 -- [origin]
  Back it up to trusted locations with git-annex copy.
```

which directly related to a recent change in git-annex.

</summary>
```
[2022-06-09 12:58:40.104765007] (Utility.Process) process [522424] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","show-ref","git-annex"]
[2022-06-09 12:58:40.107326093] (Utility.Process) process [522424] done ExitSuccess
[2022-06-09 12:58:40.107800842] (Utility.Process) process [522425] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","show-ref","--hash","refs/heads/git-annex"]
[2022-06-09 12:58:40.110214871] (Utility.Process) process [522425] done ExitSuccess
[2022-06-09 12:58:40.110383183] (Utility.Process) process [522426] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","log","refs/heads/git-annex..7fef236de205c1dedd65230dbf54b62eacb50922","--pretty=%H","-n1"]
[2022-06-09 12:58:40.112980831] (Utility.Process) process [522426] done ExitSuccess
[2022-06-09 12:58:40.117993862] (Utility.Process) process [522427] chat: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","hash-object","-w","--stdin-paths","--no-filters"]
[2022-06-09 12:58:40.1183213] (Utility.Process) process [522428] chat: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","cat-file","--batch"]
[2022-06-09 12:58:40.118508819] (Utility.Process) process [522429] feed: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","update-index","-z","--index-info"]
[2022-06-09 12:58:40.118696224] (Utility.Process) process [522430] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","diff-index","--raw","-z","-r","--no-renames","-l0","--cached","refs/heads/git-annex","--"]
[2022-06-09 12:58:40.121375579] (Utility.Process) process [522430] done ExitSuccess
[2022-06-09 12:58:40.121699292] (Utility.Process) process [522429] done ExitSuccess
[2022-06-09 12:58:40.122057043] (Annex.Branch) read remote.log
[2022-06-09 12:58:40.122743897] (Annex.Branch) read trust.log
[2022-06-09 12:58:40.122928676] (Annex.Branch) read numcopies.log
[2022-06-09 12:58:40.123825612] (Utility.Process) process [522431] chat: /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA []
[2022-06-09 12:58:40.123895201] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] <-- GETVERSION
[Level 5] Instantiating ssh manager 
[Level 5] Done importing main __init__ 
[Level 5] Importing support.network 
[Level 5] Done importing support.network 
[Level 5] Requested to provide version for cmd:git 
[Level 9] Updated GIT_DIR to /home/mih/hacking/datalad/next/tmp/myclone/.git/dl-repoannex/origin/repoannex 
[DEBUG] Run ['git', 'version'] (cwd=None) 
[Level 8] Process 522441 started 
[Level 5] STDERR: git version (Thread<(STDERR: git version, 5)>) started 
[Level 5] STDOUT: git version (Thread<(STDOUT: git version, 3)>) started 
[Level 5] process_waiter (Thread<process_waiter>) started 
[Level 5] Read 19 bytes from 522441[stdout] 
[Level 5] STDOUT: git version (Thread<(STDOUT: git version, 3)>) exiting (exit_requested: False, last data: None) 
[Level 5] process_waiter (Thread<process_waiter>) exiting 
[Level 5] STDERR: git version (Thread<(STDERR: git version, 5)>) exiting (exit_requested: False, last data: None) 
[Level 8] Process 522441 exited with return code 0 
[DEBUG] Finished ['git', 'version'] with status 0 
[Level 8] No module named 'datalad_deprecated' [gitrepo.py:<module>:3737] 
[DEBUG] Not retro-fitting GitRepo with deprecated symbols, datalad-deprecated package not found 
[DEBUG] Apply datalad-next patch to annexrepo.py:AnnexRepo.enable_remote 
[Level 5] Importing dataset 
[Level 5] Done importing dataset 
[Level 8] No module named 'requests_ftp' [http.py:<module>:62] 
[DEBUG] Failed to import requests_ftp, thus no ftp support: ModuleNotFoundError(No module named 'requests_ftp') 
[Level 5] Importing datalad.interface.results 
[Level 5] Done importing datalad.interface.results 
[DEBUG] Apply datalad-next patch to create_sibling_ghlike.py:_GitHubLike._set_request_headers 
[DEBUG] Building doc for <class 'datalad.core.local.status.Status'> 
[DEBUG] Building doc for <class 'datalad.core.local.diff.Diff'> 
[DEBUG] Building doc for <class 'datalad.core.distributed.push.Push'> 
[DEBUG] Patching datalad.core.distributed.push._transfer_data 
[DEBUG] Patching datalad.core.distributed.push.Push docstring and parameters 
[DEBUG] Building doc for <class 'datalad.core.distributed.push.Push'> 
[DEBUG] Patching datalad.support.AnnexRepo.get_export_records (new method) 
[DEBUG] Patching datalad.core.distributed.push._push 
[DEBUG] Building doc for <class 'datalad.distribution.update.Update'> 
[DEBUG] Building doc for <class 'datalad.distribution.siblings.Siblings'> 
[DEBUG] Apply datalad-next patch to siblings.py:_enable_remote 
[2022-06-09 12:58:40.452792385] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] --> VERSION 1
[2022-06-09 12:58:40.452952653] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] <-- CANVERIFY
[2022-06-09 12:58:40.453240061] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] --> CANVERIFY-YES
[2022-06-09 12:58:40.453281514] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] <-- ISSTABLE
[2022-06-09 12:58:40.453345299] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] --> ISSTABLE-NO
[2022-06-09 12:58:40.453373478] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] <-- ISCRYPTOGRAPHICALLYSECURE
[2022-06-09 12:58:40.453512855] (Annex.ExternalAddonProcess) /home/mih/env/datalad-dev/bin/git-annex-backend-XDLRA[1] --> ISCRYPTOGRAPHICALLYSECURE-NO
[2022-06-09 12:58:40.471669769] (Annex.Branch) read export.log
[2022-06-09 12:58:40.472150955] (Utility.Process) process [522450] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","diff-tree","-z","--raw","--no-renames","-l0","-r","4b825dc642cb6eb9a060e54bf8d69288fbee4904","7f0e7953e93b4c9920c2bff9534773394f3a5762","--"]
[2022-06-09 12:58:40.476549694] (Utility.Process) process [522451] chat: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","cat-file","--batch-check=%(objectname) %(objecttype) %(objectsize)"]
[2022-06-09 12:58:40.479317776] (Utility.Process) process [522452] chat: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","cat-file","--batch"]
[2022-06-09 12:58:40.481855753] (Utility.Process) process [522450] done ExitSuccess
[2022-06-09 12:58:40.486404384] (Annex.Branch) read 3f7/4a3/XDLRA--refs.log
[2022-06-09 12:58:40.48672456] (Annex.Branch) read 3f7/4a3/XDLRA--refs.log
[2022-06-09 12:58:40.487012582] (Annex.Branch) set 3f7/4a3/XDLRA--refs.log
[2022-06-09 12:58:40.48704974] (Annex.Branch) read 3f7/4a3/XDLRA--refs.log
[2022-06-09 12:58:40.487141518] (Annex.Branch) read uuid.log
  Only these untrusted locations may have copies of XDLRA--refs
        0cc19d93-2885-44c8-83ba-30c2dd3afe40 -- [origin]
  Back it up to trusted locations with git-annex copy.
[2022-06-09 12:58:40.487435721] (Annex.Branch) read activity.log
[2022-06-09 12:58:40.487713107] (Annex.Branch) set activity.log
[2022-06-09 12:58:40.488282595] (Utility.Process) process [522453] feed: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","update-index","-z","--index-info"]
[2022-06-09 12:58:40.492162749] (Utility.Process) process [522453] done ExitSuccess
[2022-06-09 12:58:40.492545997] (Utility.Process) process [522454] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","show-ref","--hash","refs/heads/git-annex"]
[2022-06-09 12:58:40.495228312] (Utility.Process) process [522454] done ExitSuccess
[2022-06-09 12:58:40.495630629] (Utility.Process) process [522455] read: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","write-tree"]
[2022-06-09 12:58:40.500163285] (Utility.Process) process [522455] done ExitSuccess
[2022-06-09 12:58:40.500551135] (Utility.Process) process [522456] chat: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","commit-tree","4c77d0557cd40fc90b148e4c37609b08862b13a2","--no-gpg-sign","-p","refs/heads/git-annex"]
[2022-06-09 12:58:40.503226831] (Utility.Process) process [522456] done ExitSuccess
[2022-06-09 12:58:40.503454986] (Utility.Process) process [522457] call: git ["--git-dir=.","--literal-pathspecs","-c","annex.debug=true","-c","annex.dotfiles=true","update-ref","refs/heads/git-annex","8ae7bfb27c860bab05cee763f7a9684d3b40840c"]
[2022-06-09 12:58:40.506732789] (Utility.Process) process [522457] done ExitSuccess
[2022-06-09 12:58:40.507288254] (Utility.Process) process [522452] done ExitSuccess
[2022-06-09 12:58:40.50747011] (Utility.Process) process [522428] done ExitSuccess
[2022-06-09 12:58:40.507759388] (Utility.Process) process [522451] done ExitSuccess
[2022-06-09 12:58:40.508061665] (Utility.Process) process [522427] done ExitSuccess
fsck: 1 failed'
```
</details>

```
% git -C myclone/.git/dl-repoannex/origin/repoannex -c diff.ignoreSubmodules=none annex fsck -f origin --fast --key XDLRA--refs -c annex.dotfiles=true
fsck XDLRA--refs 
  Only these untrusted locations may have copies of XDLRA--refs
        95652ad0-00d9-4df5-8f57-c6393772d540 -- [origin]
  Back it up to trusted locations with git-annex copy.

  (Avoid this check by running: git annex dead --key )
failed
(recording state in git...)
fsck: 1 failed
```

It is impossible to override this trust:

```
% git -C myclone/.git/dl-repoannex/origin/repoannex -c diff.ignoreSubmodules=none annex trust origin --force
trust origin 
  This remote's trust level is overridden to untrusted.
ok
(recording state in git...)
```

The recommended way to disable the check does not work:

```
% git -C myclone/.git/dl-repoannex/origin/repoannex -c diff.ignoreSubmodules=none annex dead --key XDLRA--refs                                       
dead XDLRA--refs 
git-annex: This key is still known to be present in some locations; not marking as dead.
failed
dead: 1 failed
```

**Comment by mih** [comment]
Filed https://git-annex.branchable.com/bugs/10.20220525_cannot_probe_for_untrusted_remote_key/

**Comment by joeyh** [comment]
Use git-annex checkpresentkey, that's the right tool for the job.

**Comment by mih** [comment]
Oh lord, I was blind! Thank you @joeyh !

Maybe one more question re

```
% git  annex trust origin --force
trust origin 
  This remote's trust level is overridden to untrusted.
ok
```

We are presently looking into implementing a special remote for IO with https://dataverse.org

Datasets deposited there would eventually end up archived with persistent identifiers, and, importantly, versioned at the level of an entire dataset state. In such a scenario, one would end up with a rather trustworthy storage site, even when the initial upload was performed via a git-annex export. From my point of view, it would be good to be able to reflect this trust in the git-annex setup too. Do you consider this a valid use case for overriding the override that git-annex is performing now?

Thanks again!
